src/mono/mono/mini/interp/transform.c

```diff
@@ -1098,6 +1098,30 @@ get_data_item_index (TransformData *td, void *ptr)
 {
 	gpointer p = g_hash_table_lookup (td->data_hash, ptr);
 	guint index;
+	if (p != NULL) {
+		index = GPOINTER_TO_UINT (p) - 1;
+		if (index > UINT16_MAX)
+			g_warning ("Interpreter data item index 0x%x for method '%s'  is out of range", index, td->method->name);
+		return index;
+	}
+	if (td->max_data_items == td->n_data_items) {
+		td->max_data_items = td->n_data_items == 0 ? 16 : 2 * td->max_data_items;
+		td->data_items = (gpointer*)g_realloc (td->data_items, td->max_data_items * sizeof(td->data_items [0]));
+	}
+	index = td->n_data_items;
+	td->data_items [index] = ptr;
+	++td->n_data_items;
+	g_assertf (index < (UINT16_MAX - 1), "Interpreter data item index 0x%x for method '%s' overflows", index, td->method->name);
+
+	g_hash_table_insert (td->data_hash, ptr, GUINT_TO_POINTER (index + 1));
+	return index;
+}
+
+static uint32_t
+get_data_item_wide_index (TransformData *td, void *ptr)
+{
+	gpointer p = g_hash_table_lookup (td->data_hash, ptr);
+	uint32_t index;
```

Comment:
There is quite a bit of duplicated code here. Why not just have a `get_data_item_index_wide` and the old `get_data_item_index` which calls `get_data_item_index_wide` and then asserts the index is <= G_MAXUINT16.